File: neuroops-agent/sdk/python/neuroopssdk/clients/sshclient.py
#  NeurOps Agent — Python SDK SSH Client
#  Copyright (c) NeurOps 2025. All rights reserved.
#
#  Wraps Paramiko for metric and runbook plugins that collect data via SSH.
#  Supports password auth, key-based auth, SFTP file transfer, and
#  interactive (pty) sessions for multi-step CLI interactions.
#
#  Usage:
#    client = SSHClient(context, logger)
#    result = client.connect()
#    if result["status"] == "succeed":
#        out = client.execute("df -h")
#        client.disconnect()

import io
import time

import paramiko

from neuroopssdk.constants import Constant


class SSHClient:

    DEFAULT_PORT    = 22
    DEFAULT_TIMEOUT = 30

    def __init__(self, context: dict, logger):
        self._log      = logger
        self._host     = context.get(Constant.OBJECT_IP, "")
        self._port     = int(context.get("ssh.port", context.get(Constant.PORT, self.DEFAULT_PORT)))
        self._username = context.get(Constant.USERNAME, "")
        self._password = context.get(Constant.PASSWORD, "")
        self._key      = context.get("ssh.key", "")
        self._passphrase = context.get("ssh.passphrase", "")
        self._timeout  = int(context.get(Constant.TIMEOUT, self.DEFAULT_TIMEOUT))
        self._client: paramiko.SSHClient | None = None

    # ── Lifecycle ──────────────────────────────────────────────────────────

    def connect(self) -> dict:
        try:
            self._client = paramiko.SSHClient()
            self._client.set_missing_host_key_policy(paramiko.AutoAddPolicy())

            kwargs = dict(
                hostname=self._host,
                port=self._port,
                username=self._username,
                timeout=self._timeout,
                look_for_keys=False,
                allow_agent=False,
            )

            if self._key:
                pkey = self._load_key(self._key, self._passphrase)
                if pkey is None:
                    return self._fail(
                        f"Failed to load SSH private key for {self._host}",
                        Constant.ERROR_INVALID_CREDENTIALS,
                    )
                kwargs["pkey"] = pkey
            else:
                kwargs["password"] = self._password

            self._client.connect(**kwargs)
            self._log.infof(Constant.INFO_MSG_CONNECTED, self._host, self._port)
            return {"status": Constant.STATUS_SUCCEED}

        except paramiko.AuthenticationException:
            return self._fail(
                f"Authentication Error: Invalid credentials for {self._host}:{self._port}",
                Constant.ERROR_INVALID_CREDENTIALS,
            )
        except paramiko.ssh_exception.NoValidConnectionsError:
            return self._fail(
                f"Connection Error: Could not connect to {self._host}:{self._port}",
                Constant.ERROR_CONNECTION_FAILED,
            )
        except TimeoutError:
            return self._fail(
                f"Timeout Error: SSH connection to {self._host}:{self._port} timed out",
                Constant.ERROR_TIMEOUT,
            )
        except Exception as e:
            return self._fail(str(e), Constant.ERROR_INTERNAL)

    def disconnect(self):
        if self._client:
            try:
                self._client.close()
            except Exception:
                pass
            self._client = None
            self._log.infof("Disconnected from %s:%d", self._host, self._port)

    # ── Command execution ──────────────────────────────────────────────────

    def execute(self, command: str, timeout: int = None) -> dict:
        """
        Run a single command and return stdout/stderr.
        Returns: {"status": "succeed", "output": "...", "error_output": "..."}
        """
        if not self._client:
            return self._fail("Not connected", Constant.ERROR_CONNECTION_FAILED)
        try:
            t = timeout or self._timeout
            _, stdout, stderr = self._client.exec_command(command, timeout=t)
            out = stdout.read().decode(errors="replace").strip()
            err = stderr.read().decode(errors="replace").strip()
            self._log.tracef("SSH command: %s → %s", command, out[:200])
            return {
                "status":       Constant.STATUS_SUCCEED,
                "output":       out,
                "error_output": err,
            }
        except Exception as e:
            return self._fail(
                f"Command execution failed [{command}]: {e}",
                Constant.ERROR_COMMAND_EXECUTION_FAILED,
            )

    def execute_many(self, commands: list) -> dict:
        """
        Execute multiple commands sequentially.
        Returns: {"status": "succeed", "results": [<per-cmd result>, ...]}
        """
        results = []
        for cmd in commands:
            r = self.execute(cmd)
            results.append({"command": cmd, **r})
            if r["status"] != Constant.STATUS_SUCCEED:
                break
        return {
            "status":  Constant.STATUS_SUCCEED,
            "results": results,
        }

    # ── Interactive (PTY) session ──────────────────────────────────────────

    def open_shell(self, width: int = 200, height: int = 50) -> paramiko.Channel:
        """Open an interactive PTY channel (for multi-step CLI workflows)."""
        if not self._client:
            raise RuntimeError("Not connected")
        channel = self._client.invoke_shell(width=width, height=height)
        channel.settimeout(self._timeout)
        time.sleep(1)
        self._flush(channel)
        return channel

    def send_command(self, channel: paramiko.Channel, command: str,
                     wait_pattern: str = None, delay: float = 1.0) -> str:
        """
        Send a command on an open shell channel and collect output.
        Optionally waits until wait_pattern appears in the output.
        """
        channel.send(command + "\n")
        time.sleep(delay)
        return self._read_until(channel, wait_pattern, timeout=self._timeout)

    # ── SFTP ──────────────────────────────────────────────────────────────

    def upload(self, local_path: str, remote_path: str) -> dict:
        if not self._client:
            return self._fail("Not connected", Constant.ERROR_CONNECTION_FAILED)
        try:
            sftp = self._client.open_sftp()
            sftp.put(local_path, remote_path)
            sftp.close()
            return {"status": Constant.STATUS_SUCCEED}
        except Exception as e:
            return self._fail(f"SFTP upload failed: {e}", Constant.ERROR_INTERNAL)

    def download(self, remote_path: str, local_path: str) -> dict:
        if not self._client:
            return self._fail("Not connected", Constant.ERROR_CONNECTION_FAILED)
        try:
            sftp = self._client.open_sftp()
            sftp.get(remote_path, local_path)
            sftp.close()
            return {"status": Constant.STATUS_SUCCEED}
        except Exception as e:
            return self._fail(f"SFTP download failed: {e}", Constant.ERROR_INTERNAL)

    def read_remote_file(self, remote_path: str) -> dict:
        """Read a remote file and return its contents as a string."""
        if not self._client:
            return self._fail("Not connected", Constant.ERROR_CONNECTION_FAILED)
        try:
            sftp = self._client.open_sftp()
            with sftp.open(remote_path, "r") as f:
                content = f.read().decode(errors="replace")
            sftp.close()
            return {"status": Constant.STATUS_SUCCEED, "content": content}
        except Exception as e:
            return self._fail(f"Failed to read {remote_path}: {e}", Constant.ERROR_INTERNAL)

    # ── Helpers ────────────────────────────────────────────────────────────

    def _load_key(self, key_data: str, passphrase: str = None):
        passphrase = passphrase.encode() if passphrase else None
        for key_class in (
            paramiko.RSAKey,
            paramiko.Ed25519Key,
            paramiko.ECDSAKey,
            paramiko.DSSKey,
        ):
            try:
                return key_class.from_private_key(io.StringIO(key_data), password=passphrase)
            except Exception:
                continue
        return None

    def _flush(self, channel: paramiko.Channel) -> str:
        buf = b""
        while channel.recv_ready():
            buf += channel.recv(4096)
        return buf.decode(errors="replace")

    def _read_until(self, channel: paramiko.Channel,
                    pattern: str = None, timeout: float = 30) -> str:
        import time as _time
        buf = ""
        start = _time.time()
        while True:
            if channel.recv_ready():
                buf += channel.recv(4096).decode(errors="replace")
                if pattern and pattern in buf:
                    break
            elif _time.time() - start > timeout:
                break
            else:
                _time.sleep(0.1)
        return buf

    def _fail(self, message: str, code: str = Constant.ERROR_INTERNAL) -> dict:
        self._log.error(message)
        return {
            "status":     Constant.STATUS_FAIL,
            "error":      message,
            "error.code": code,
        }

    # Constant shorthands (avoid AttributeError for missing consts)
    class _C:
        INFO_MSG_CONNECTED = "Connected to %s:%d"

    INFO_MSG_CONNECTED = "Connected to %s:%d"

File: neuroops-agent/sdk/go/clients/ssh/sshclient.go
/*
 * NeurOps Agent — Go SDK SSH Client
 * Copyright (c) NeurOps 2025. All rights reserved.
 *
 * Wraps golang.org/x/crypto/ssh.  Used by Go metric and runbook plugins
 * that collect data or run remediation on remote Linux/Unix hosts.
 *
 * Supports: password auth, private-key auth (RSA, ECDSA, Ed25519),
 *           command execution, SFTP file transfer, interactive PTY shells.
 *
 * Usage:
 *   client := ssh.New(context, logger)
 *   if err := client.Connect(); err != nil { ... }
 *   defer client.Disconnect()
 *   out, err := client.Run("df -h")
 */

package ssh

import (
	"fmt"
	"net"
	"strings"
	"time"

	"golang.org/x/crypto/ssh"

	"github.com/neuroops/agent/internal/logger"
	"github.com/neuroops/agent/sdk/go/types"
)

// Client wraps an SSH connection for plugin use.
type Client struct {
	host     string
	port     int
	username string
	password string
	keyData  string
	timeout  time.Duration
	log      *logger.Logger
	conn     *ssh.Client
}

// New creates an SSH Client from a plugin context map.
func New(ctx types.NeurOpsMap, log *logger.Logger) *Client {
	port := ctx.GetInt("ssh.port")
	if port == 0 {
		port = ctx.GetInt("port")
	}
	if port == 0 {
		port = 22
	}

	timeout := ctx.GetInt("timeout")
	if timeout == 0 {
		timeout = 30
	}

	return &Client{
		host:     ctx.GetString("object.ip"),
		port:     int(port),
		username: ctx.GetString("username"),
		password: ctx.GetString("password"),
		keyData:  ctx.GetString("ssh.key"),
		timeout:  time.Duration(timeout) * time.Second,
		log:      log,
	}
}

// Connect establishes the SSH connection.
func (c *Client) Connect() error {
	var authMethods []ssh.AuthMethod

	if c.keyData != "" {
		signer, err := parseKey(c.keyData, "")
		if err != nil {
			return fmt.Errorf("invalid SSH key: %w", err)
		}
		authMethods = append(authMethods, ssh.PublicKeys(signer))
	}
	if c.password != "" {
		authMethods = append(authMethods, ssh.Password(c.password))
	}
	if len(authMethods) == 0 {
		return fmt.Errorf("no SSH auth method configured (password or ssh.key required)")
	}

	cfg := &ssh.ClientConfig{
		User:            c.username,
		Auth:            authMethods,
		HostKeyCallback: ssh.InsecureIgnoreHostKey(),
		Timeout:         c.timeout,
	}

	addr := fmt.Sprintf("%s:%d", c.host, c.port)
	conn, err := net.DialTimeout("tcp", addr, c.timeout)
	if err != nil {
		return fmt.Errorf("TCP dial %s failed: %w", addr, err)
	}

	sshConn, chans, reqs, err := ssh.NewClientConn(conn, addr, cfg)
	if err != nil {
		return fmt.Errorf("SSH handshake with %s failed: %w", addr, err)
	}

	c.conn = ssh.NewClient(sshConn, chans, reqs)
	c.log.Infof("SSH connected to %s:%d", c.host, c.port)
	return nil
}

// Disconnect closes the SSH connection.
func (c *Client) Disconnect() {
	if c.conn != nil {
		_ = c.conn.Close()
		c.conn = nil
	}
}

// Run executes a single command and returns stdout output.
func (c *Client) Run(command string) (string, error) {
	if c.conn == nil {
		return "", fmt.Errorf("not connected")
	}
	session, err := c.conn.NewSession()
	if err != nil {
		return "", fmt.Errorf("creating SSH session: %w", err)
	}
	defer session.Close()

	out, err := session.CombinedOutput(command)
	if err != nil {
		return strings.TrimSpace(string(out)), fmt.Errorf("command %q: %w", command, err)
	}
	return strings.TrimSpace(string(out)), nil
}

// RunMany executes multiple commands and returns their outputs.
// Stops on the first error.
func (c *Client) RunMany(commands []string) (map[string]string, error) {
	results := make(map[string]string, len(commands))
	for _, cmd := range commands {
		out, err := c.Run(cmd)
		if err != nil {
			return results, fmt.Errorf("command %q failed: %w", cmd, err)
		}
		results[cmd] = out
	}
	return results, nil
}

// ── Key parsing ───────────────────────────────────────────────────────────────

func parseKey(keyData, passphrase string) (ssh.Signer, error) {
	pemBytes := []byte(keyData)
	if passphrase != "" {
		return ssh.ParsePrivateKeyWithPassphrase(pemBytes, []byte(passphrase))
	}
	return ssh.ParsePrivateKey(pemBytes)
}
